### PR TITLE
Documentation about upgradable contracts

### PR DESCRIPTION
**Pull request by xgreenx**
[Preview](https://github.com/Supercolony-net/openbrush-contracts/blob/feature/upgradable-contract/docs/docs/smart-contracts/upgradeable.md)

### CODE REVIEW

File: examples/payment_splitter/README.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ accounts but kept in this contract, and the actual transfer is triggered as a se
 function. `release` pays out to only the provided address. If you will have many people to pay out, especially if often, you will likely
 want to use the `releaseAll` method instead to save you a lot of time.
 
-** Note **: In the substrate balance of contract decreases each block. Because it pays rent for the storage.
+**Note**: In the substrate balance of contract decreases each block. Because it pays rent for the storage.
```
Comment (PierreOssun) [inline examples/payment_splitter/README.md:15]: I guess this was removed =)

Comment (xgreenx) [reply examples/payment_splitter/README.md:15]: Good catch!=)

File: docs/docs/smart-contracts/upgradeable.md [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+should be able to call that function.
+
+All suggestions described above are applicable for that kind of upgradeable contracts. 
+Better to have an upgradeable storage layout, initialization function for new versions 
+of your contract, permission system, etc.
+
+### The `Diamond` Standard
+
+Using `Diamond` Standard you can add support for several facets(logic layers) that 
+can be upgraded. [That standard](https://eips.ethereum.org/EIPS/eip-2535) came
```
Comment (coreggon11) [inline docs/docs/smart-contracts/upgradeable.md:403]:
```suggestion
can be upgraded. [This standard](https://eips.ethereum.org/EIPS/eip-2535) came 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+pub struct OwnableData {
+    // Owner of the contract
+    pub owner: AccountId,
+    // Reserved fields for future upgrades
+    pub _reserved: Option<()>,
+}
+```
+
+`PSP22Data` and `OwnableData` have their storage keys. Both contain additional fields, 
+unrelated to business logic, `_reserved` for future upgrades(it adds overhead in one byte).
```
Comment (coreggon11) [inline docs/docs/smart-contracts/upgradeable.md:470]:
```suggestion
`PSP22Data` and `OwnableData` have their storage keys. Both contain an additional field 
unrelated to business logic `_reserved` for future upgrades(it adds overhead in one byte).
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+        #[ink(message)]
+        pub fn init_with_supply(&mut self, total_supply: Balance) {
+            assert_eq!(Self::env().caller(), self.ownable.owner, "Only owner can init contract");
+            ...
+        }
+    }
+}
+```
+
+You can deploy the code of the facet into the blockchain. After, you can
```
Comment (coreggon11) [inline docs/docs/smart-contracts/upgradeable.md:519]:
```suggestion
You can deploy the code of the facet to the blockchain. After that, you can 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+# Upgradeable contract
+
+## Overview
+
+Smart contracts are immutable by default, adding a layer of security and trust to the 
+contracts. But software quality depends on the ability to upgrade source code to 
+produce iterative releases. A certain degree of mutability is needed for bug fixing 
+and potential product improvements.
+
+Upgradeability allows for experimenting and deploying the product at the early stage,
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:15]:
```suggestion
Upgradeability allows experimenting and deploying the product at the early stage, 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 6]
```diff
@@ -0,0 +1,693 @@
...
+
+Smart contracts are immutable by default, adding a layer of security and trust to the 
+contracts. But software quality depends on the ability to upgrade source code to 
+produce iterative releases. A certain degree of mutability is needed for bug fixing 
+and potential product improvements.
+
+Upgradeability allows for experimenting and deploying the product at the early stage, 
+always leaving the chance to fix vulnerabilities and progressively add features. 
+It is more actual right now while ink! and contract-pallet under active development. 
+Upgradeable contracts are not a Bug if they are developed consciously with
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:18]:
```suggestion
Upgradeable contracts are not a bug if they are developed consciously with 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 7]
```diff
@@ -0,0 +1,693 @@
...
+![](assets/20220719_075309_F016550A-65D2-4DCD-A60A-D1A70B38D813.jpeg)
+
+During compilation ink! inserts code to work with storage and ink! knows how to store 
+each type in which storage cell. How exactly it works is not a part of that tutorial. 
+The main point is that each type knows how to operate with each field and operate with 
+storage, cause of a unique identifier. In the old version of ink! the identifier is 
+`[u8; 32]` in a [new version](https://github.com/paritytech/ink/issues/1134) it is `u32`.
+
+So, each data is stored under its unique identifier - the storage key. The value of the 
+key is the sequence of bytes - serialized(by SCALE codec) data type. The logic layer
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:44]:
```suggestion
key is the sequence of bytes - serialized (by SCALE codec) data type. The logic layer 
```

File: docs/docs/smart-contracts/upgradeable.md (hunk 6 above)
Comment (xgreenx) [reply docs/docs/smart-contracts/upgradeable.md:18]: I tried to highlight it with capital letter=)

File: docs/docs/smart-contracts/upgradeable.md [hunk 8]
```diff
@@ -0,0 +1,693 @@
...
+### Rules for upgradeable storage layout
+
+The contract can have several logic layers (like in the `Diamond` pattern). 
+So we will define rules in terms of several layers, but it is also applicable 
+for upgradeable contracts with the `Proxy` pattern or `set_code_hash`.
+
+1. The data stored under the storage key should use the same serialization and 
+deserialization mechanism across all logic layers. Otherwise, some logic layers will not 
+be able to deserialize the data type and fail.
+1. Each logic unit(in most cases, it is a type) should occupy the same storage key across
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:60]:
```suggestion
1. Each logic unit (in most cases, it is a type) should occupy the same storage key across
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 9]
```diff
@@ -0,0 +1,693 @@
...
+
+The manual setting of the storage key for each field allows following the rules but 
+makes development harder. ink! allows you to manually implement all utility traits and 
+specify the storage key you want to use for each field. If your contract has 20 fields, 
+then you need to set 20 storage keys.
+
+The main suggestion is to design your storage layout as a bunch of logic units and 
+assign a unique storage key per logic unit. The logic unit can be one field or a bunch 
+of fields. In the scope of the logic unit, you can use automatically calculated keys 
+offsetted with the storage key of the logic unit, or you can use the same approach
```
Comment (Artemka374) [inline docs/docs/smart-contracts/upgradeable.md:90]:
```suggestion
offset with the storage key of the logic unit, or you can use the same approach 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+For example, for an upgradeable `PSP22` contract, you can add the `init_with_supply` method:
+
+```rust
+#[ink(message)]
+pub fn init_with_supply(&mut self, total_supply: Balance) -> Result<(), PSP22Error> {
+    self._mint(Self::env().caller(), total_supply)
+}
+```
+
+You can add as many methods as you wish with any arguments(the same as constructors).
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:267]:
```suggestion
You can add as many methods as you wish with any arguments (the same as constructors).
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 11]
```diff
@@ -0,0 +1,693 @@
...
+##### Remark about logic units and OpenBrush
+
+All data structures for contracts provided by OpenBrush are upgradeable logic units. 
+So contracts support upgradeability by default.
+
+You can access the default implementation when you embed OpenBrush data structures 
+into your contract. You can use default implementation internally, or you can make 
+it public.
+
+For example, when you embed `psp22::Data` into your `Contract`. You can already
```
Comment (Artemka374) [inline docs/docs/smart-contracts/upgradeable.md:619]:
```suggestion
For example, when you embed `psp22::Data` into your `Contract`, you can already 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 12]
```diff
@@ -0,0 +1,693 @@
...
+This is the illustration of the flow of the `Diamond` pattern:
+
+![](assets/20220715_130335_47FD0F8D-60F3-4FDF-82F4-672402FDC5D1.jpeg)
+
+Each method in the smart contract has a selector. It is used as an identifier during the 
+smart contract call to execute the right logic.
+Each facet has a list of selectors that describe which methods are supported by the facet. 
+Each selector is unique and belongs only to one facet. So selectors of facets can't overlap. 
+`Diamond` contract knows which facet is responsible for which selector and forwards each 
+call to the smart contract to the corresponding facet(logic layer).
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:416]:
```suggestion
call to the smart contract to the corresponding facet (logic layer).
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 13]
```diff
@@ -0,0 +1,693 @@
...
+`Diamond` contract has a function `diamond_cut` that allows registering each facet.
+
+OpenBrush provides default implementation for `Diamond` standard on ink!.
+For more details you can check [Diamond](diamond.md).
+
+All suggestions above ideally describe how to develop an upgradeable contract
+with multi-logic layers and many logic units.
+So here we will describe how to write facets(logic layers) with OpenBrush.
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:424]:
```suggestion
So here we will describe how to write facets (logic layers) with OpenBrush.
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 14]
```diff
@@ -0,0 +1,693 @@
...
+
+#### Logic units for facet
+
+Each facet(logic layer) can have zero, one, or many logic units that work
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:428]:
```suggestion
Each facet (logic layer) can have zero, one, or many logic units that work 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 15]
```diff
@@ -0,0 +1,693 @@
...
+All suggestions above ideally describe how to develop an upgradeable contract
+with multi-logic layers and many logic units.
+So here we will describe how to write facets(logic layers) with OpenBrush.
+
+#### Logic units for facet
+
+Each facet(logic layer) can have zero, one, or many logic units that work 
+with storage.
+
+Each logic unit should have a unique storage key and be upgradeable(support initialization
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:431]:
```suggestion
Each logic unit should have a unique storage key and be upgradeable (support initialization
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 16]
```diff
@@ -0,0 +1,693 @@
...
+on demand, use storage key as an offset for all inner fields). It can be a struct with 
+one or many fields(structs without fields are useless) or an enum with at least two
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:433]:
```suggestion
one or many fields (structs without fields are useless) or an enum with at least two 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 17]
```diff
@@ -0,0 +1,693 @@
...
+
+#### Logic units for facet
+
+Each facet(logic layer) can have zero, one, or many logic units that work 
+with storage.
+
+Each logic unit should have a unique storage key and be upgradeable(support initialization
+on demand, use storage key as an offset for all inner fields). It can be a struct with 
+one or many fields(structs without fields are useless) or an enum with at least two 
+variants(an enum with one variant is a structure). You can define struct/enum with the
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:434]:
```suggestion
variants (an enum with one variant is a structure). You can define struct/enum with the 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 18]
```diff
@@ -0,0 +1,693 @@
...
+}
+```
+
+You can deploy the code of the facet to the blockchain. After that, you can 
+register your facet in `Diamond` via the `diamond_cut` method or in `Proxy` via 
+`change_delegate_call` method.
+
+#### Interaction between facets
+
+During development, you can have cases when one logic layer(for example `FacetA`) needs to interact with
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:525]:
```suggestion
During development, you can have cases when one logic layer (for example `FacetA`) needs to interact with 
```

File: docs/docs/smart-contracts/upgradeable.md [hunk 19]
```diff
@@ -0,0 +1,693 @@
...
+```
+
+You can deploy the code of the facet to the blockchain. After that, you can 
+register your facet in `Diamond` via the `diamond_cut` method or in `Proxy` via 
+`change_delegate_call` method.
+
+#### Interaction between facets
+
+During development, you can have cases when one logic layer(for example `FacetA`) needs to interact with 
+another logic unit or logic layer(for example `FacetB`).
```
Comment (o-tsaruk) [inline docs/docs/smart-contracts/upgradeable.md:526]:
```suggestion
another logic unit or logic layer (for example `FacetB`). 
```

**Review by o-tsaruk — APPROVED** [review]

**Review by Artemka374 — APPROVED** [review]

**Pull request merged**